### ISSUE TITLE
EQL: Filter pipe semantics

### ISSUE THREAD

**Issue by costin** [elastic/elasticsearch#61854]
With the support for filter pipe in EQL, it is worth discussing its semantics since there seems to be a gap between the assumptions on how it should work and existing tests.

To recap, take a look at the filter description in the [EQL Python](https://eql.readthedocs.io/en/latest/query-guide/pipes.html#filter):

> The filter pipe will only output events that match the criteria. With simple queries, this can be accomplished by adding and to the search criteria. It’s most commonly used to filter sequences or with other pipes.

To wit:

```
network where true
| filter id > 10
```

is the same as

```
network where true and id > 10
```

For sequences:

```
sequence 
[process where a > 1]
[network where b < 10]
| filter c == 5
```

this becomes

```
sequence 
[process where a > 1 and c == 5]
[network where b < 10 and c == 5]
```

This is my interpretation, how filter pipes work in EQL in ES and I believe how SIEM thinks about it as well.

However that doesn't seem to be the way EQL in Python works based on the existing tests.
It seems to applies the filter in declaration order on the results of a sequence or query.
That is, instead of filtering events *before* being matched, it filters *after* they are matched.

Assume the following stream of events A1, A2 and B3, in this order.

```
sequence
[process where A]
[process where B]
| filter A != A2
```

If the filter is applied before matching, A2 is eliminated and the resulting sequence is [A1,B3].
If the filter is applied after matching, the sequence produces [A2, B3] which then gets filtered out.

As a tangent, this affects the way other pipes get applied as well:

```
process where true
| head 10
| filter id > 10
```

Assume the id is monotonically increasing and it starts from 1. Should the result be:
[] (head 10, returns process with id 1, ... 10] which then get filtered out or
[11, 12 ... 20] ?

In other words are the following queries equivalent:
 
```
process where true
| head 10
| filter id > 10
```

```
process where true
| filter id > 10
| head 10
```
```
process where id > 10
| head 10
```

**Comment by rw-access** [comment]
The order of pipes defines the semantics and they are written in the expected order of execution. Of course, we may rearrange them at execution time when we can guarantee that we preserve semantics.

> In other words are the following queries equivalent:
* Queries 2 and 3 are identical.
    * *find the first 10 process events, and of those, only show me those with id > 10*
* Query 1 will most likely not match the same results
    * *find the process events with id > 10, and only return the first 10 of those*

**Comment by costin** [comment]
If that behavior is post matching/declaration order, this results in excluding sequences because the matching events don't match a filter since:

**Comment by rw-access** [comment]
@costin if you specify a `| filter` for a `sequence`, whenever you refer to a field, you need to mention which event you're referring to. otherwise it defaults to the first.


```
sequence
  [process where A]
  [file where B]
| filter foo == "bar"
```

is identical to 
```
sequence
  [process where A]
  [file where B]
| filter events[0].foo == "bar"
```

`sequence ... | filter` means that you're filtering on _complete_ sequences. The criteria in the filter has the full sequence in context, instead of specifying a condition that is ANDed to each term.

**Comment by costin** [comment]
> if you specify a | filter for a sequence, whenever you refer to a field, you need to mention which event you're referring to. otherwise it defaults to the first (events[0].).

Elasticsearch EQL does not currently support array notation nor pseudo-fields (e.g. `events`).

As the filter pipe (in its initial form) is out of scope for 7.10 and not currently on the roadmap, I'm closing this issue.
